modules/api_loader.lua

```diff
@@ -9,73 +9,148 @@ local function get_name(xml_element)
     end
     return name
 end
+ 
 
-local function loadEnums(api, dest)
-  local enums = api:xpath("//interface/enum")
-  for _, e in ipairs(enums) do
-    local enum = { }
-    local i = 1
-    for _, item in ipairs(e:children("element")) do
-      enum[item:attr("name")] = i
-      i = i + 1
+local function LoadResultCodes( param )
+  local resultCodes ={}
+  local i = 1
+  for _, item in ipairs(param:children("element")) do
+     local name = item:attr("name")
+     resultCodes[i]=name
+     i=i + 1
     end
-    dest.enum[get_name(e)] = enum
+  return resultCodes
+end
+
+local function LoadParamsInFunction(param, interface)
+  local name = param:attr("name")
+  local p_type = param:attr("type")
+  local minlength = tonumber(param:attr("minlength"))
+  local maxlength = tonumber(param:attr("maxlength"))
+  local minsize =  tonumber(param:attr("minsize"))
+  local maxsize = tonumber(param:attr("maxsize"))
+  local minvalue = tonumber(param:attr("minvalue"))
+  local maxvalue = tonumber(param:attr("maxvalue"))
+  local mandatory = param:attr("mandatory")
+  local array = param:attr("array")
+
+  if mandatory == nil then 
+    mandatory = true
+  end
+
+  if array == nil then 
+    array = false
   end
+
+  local result_codes = nil
+  if name == "resultCode" and p_type == "Result" then 
+    result_codes  = LoadResultCodes(param) 
+  end
+
+  local data = {}
+  data["type"]=p_type
+  data["mandatory"]= mandatory
+  data["array"] = array
+  data["minlength"] = minlength
+  data["maxlength"] = maxlength
+  data["minsize"] = minsize
+  data["maxsize"] = maxsize
+  data["minvalue"] = minvalue
+  data["maxvalue"] = maxvalue
+  data["resultCodes"] = result_codes
+  return name, data
 end
 
-local function loadStructs(api, dest)
-  local structs = api:xpath("//interface/struct")
-  for _, s in ipairs(structs) do
-    local struct = { }
-    for _, item in ipairs(s:children("param")) do
-      struct[item:attr("name")] = item:attributes()
+
+
+ local function LoadEnums(api, dest)
+   for first, v in pairs (dest.interface) do
+    for _, s in ipairs(v.body:children("enum")) do
+      local name = s:attr("name")
+      dest.interface[first].enum[name]={}
+      local i = 1
+      for _,e in ipairs(s:children("element")) do
+        local enum_value = e:attr("name")
+        dest.interface[first].enum[name][enum_value]=i
+        i= i + 1
+      end
     end
-    dest.struct[get_name(s)] = struct
   end
+ end
+ 
+ local function LoadStructs(api, dest)
+   for first, v in pairs (dest.interface) do
+    for _, s in ipairs(v.body:children("struct")) do
+      local name = s:attr("name")
+      local temp_param = {}
+      local temp_func = {}
+      temp_func["name"] = name
```

Comment:
I see such a lot of the common code as following:
```
local value = in["value"];
out["value"] = value;
```
Will it be more clear and readable use it like 
```
SetValue(from, to, "value") ?
```
It could reduce a lot of code here and also guaranty `"value"` string correctness.